File: python/example_code/iam/create_policy_version.py
# snippet-comment:[These are tags for the AWS doc team's sample catalog. Do not remove.]
# snippet-sourcedescription:[create_policy_version.py demonstrates how to create a new version of an IAM policy.]
# snippet-service:[iam]
# snippet-keyword:[AWS Identity and Access Management (IAM)]
# snippet-keyword:[Python]
# snippet-sourcesyntax:[python]
# snippet-keyword:[Code Sample]
# snippet-sourcetype:[snippet]
# snippet-sourcedate:[2019-04-10]
# snippet-sourceauthor:[AWS]

# Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
# http://aws.amazon.com/apache2.0/
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.

# snippet-start:[iam.python.create_policy_version.complete]

import json
import boto3
from botocore.exceptions import ClientError

# Set this value before running the program
# Policy to create a new version of
policy_arn = 'arn:aws:iam::123456789012:policy/POLICY_NAME'

# Define the new version of the policy
new_policy = {
    'Version': '2012-10-17',
    'Statement': [
        {
            'Sid': 'Statement1',
            'Effect': 'Allow',
            'Action': 'EC2:*',
            'Resource': '*'
        },
        {
            'Sid': 'Statement2',
            'Effect': 'Allow',
            'Action': 'S3:*',
            'Resource': '*'
        }
    ]
}
print(f'Creating new version of IAM policy {policy_arn}')
print(json.dumps(new_policy))

# Create the new version of the policy and set it as the default version
try:
    iam = boto3.client('iam')
    response = iam.create_policy_version(PolicyArn=policy_arn,
                                         PolicyDocument=json.dumps(new_policy),
                                         SetAsDefault=True)

    print(f'Policy Version Created: {response["PolicyVersion"]["VersionId"]}')
except ClientError as e:
    print(e)

# snippet-end:[iam.python.create_policy_version.complete]

File: php/example_code/iam/CreatePolicyVersion.php
<?php
/**
 * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * This file is licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License. A copy of
 * the License is located at
 *
 * http://aws.amazon.com/apache2.0/
 *
 * This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
require 'vendor/autoload.php';

use Aws\Iam\IamClient;
use Aws\Exception\AwsException;

/**
 * Creates a new version of the specified managed policy.
 *
 * This code expects that you have AWS credentials set up per:
 * https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_credentials.html
 */

$client = new IamClient([
    'profile' => 'default',
    'region' => 'us-west-2',
    'version' => '2010-05-08'
]);

$myManagedPolicy = '{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "logs:CreateLogGroup",
            "Resource": "RESOURCE_ARN"
        },
        {
            "Effect": "Allow",
            "Action": [
            "dynamodb:DeleteItem",
            "dynamodb:GetItem",
            "dynamodb:PutItem",
            "dynamodb:Scan",
            "dynamodb:UpdateItem"
        ],
            "Resource": "RESOURCE_ARN"
        }
    ]
}';

try {
    $result = $client->createPolicyVersion(array(
        // PolicyArn is required
        'PolicyArn' => 'POLICY_ARN',
        // PolicyDocument is required
        'PolicyDocument' => $myManagedPolicy,
        'SetAsDefault' => true
    ));
    var_dump($result);
} catch (AwsException $e) {
    // output error message if fails
    error_log($e->getMessage());
}
 

//snippet-sourcedescription:[CreatePolicyVersion.php demonstrates how to create a new version of the specified managed policy.]
//snippet-keyword:[PHP]
//snippet-keyword:[AWS SDK for PHP v3]
//snippet-keyword:[Code Sample]
//snippet-keyword:[AWS Identity and Access Management (IAM)]
//snippet-service:[iam]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[2018-09-20]
//snippet-sourceauthor:[jschwarzwalder]

